packages/dependency-path/src/index.ts

```diff
@@ -104,13 +108,22 @@ export function parse (dependencyPath: string) {
   const name = parts[0].startsWith('@')
     ? `${parts.shift()}/${parts.shift()}` // eslint-disable-line @typescript-eslint/restrict-template-expressions
     : parts.shift()
-  let version = parts.shift()
+  let version = parts.join('/')
   if (version) {
-    const underscoreIndex = version.indexOf('_')
+    let peerSepIndex!: number
     let peersSuffix: string | undefined
-    if (underscoreIndex !== -1) {
-      peersSuffix = version.substring(underscoreIndex + 1)
-      version = version.substring(0, underscoreIndex)
+    if (version.includes('(') && version.endsWith(')')) {
+      peerSepIndex = version.indexOf('(')
+      if (peerSepIndex !== -1) {
+        peersSuffix = version.substring(peerSepIndex)
+        version = version.substring(0, peerSepIndex)
+      }
```

Comment:
@zkochan This does not work for patched dependencies. `_` is used to separate version and patch hash: `/ts-node/10.9.1_k5brw22k7hadiw3hedogf4eiee(@swc/core@1.3.27)(@types/node@18.11.18)(typescript@4.9.4)`

I believe this should fix it, but I am not familiar with the codebase to know if it might affect anything else:
```suggestion
      peerSepIndex = version.indexOf('(')
      if (peerSepIndex !== -1) {
        peersSuffix = version.substring(peerSepIndex)
        version = version.substring(0, peerSepIndex)
      }

      let patchSepIndex = version.indexOf("_");
      if(patchSepIndex !== -1) {
        version = version.substring(0, patchSepIndex);
      }
```

Reply:
My suggestion didn't quite work either. Here is a PR that fixes the isssue: https://github.com/pnpm/pnpm/pull/5964